exporters/trace/jaeger/src/main/java/io/opencensus/exporter/trace/jaeger/JaegerExporterHandler.java

```diff
@@ -0,0 +1,273 @@
+
+import com.google.common.collect.Lists;
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
+import com.uber.jaeger.senders.HttpSender;
+import com.uber.jaeger.thriftjava.Log;
+import com.uber.jaeger.thriftjava.Process;
+import com.uber.jaeger.thriftjava.Span;
+import com.uber.jaeger.thriftjava.SpanRef;
+import com.uber.jaeger.thriftjava.SpanRefType;
+import com.uber.jaeger.thriftjava.Tag;
+import com.uber.jaeger.thriftjava.TagType;
+import io.opencensus.common.Function;
+import io.opencensus.common.Timestamp;
+import io.opencensus.trace.Annotation;
+import io.opencensus.trace.AttributeValue;
+import io.opencensus.trace.Link;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.SpanId;
+import io.opencensus.trace.Status;
+import io.opencensus.trace.TraceId;
+import io.opencensus.trace.TraceOptions;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import io.opencensus.trace.export.SpanData;
+import io.opencensus.trace.export.SpanExporter;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.thrift.TException;
+
+@NotThreadSafe
+final class JaegerExporterHandler extends SpanExporter.Handler {
+  private static final Logger logger = Logger.getLogger(JaegerExporterHandler.class.getName());
+  private static final Tracer tracer = Tracing.getTracer();
+
+  private static final Function<? super String, Tag> stringAttributeConverter =
+      new Function<String, Tag>() {
+        @Override
+        public Tag apply(final String value) {
+          final Tag tag = new Tag();
+          tag.setVType(TagType.STRING);
+          tag.setVStr(value);
+          return tag;
+        }
+      };
+
+  private static final Function<? super Boolean, Tag> booleanAttributeConverter =
+      new Function<Boolean, Tag>() {
+        @Override
+        public Tag apply(final Boolean value) {
+          final Tag tag = new Tag();
+          tag.setVType(TagType.BOOL);
+          tag.setVBool(value);
+          return tag;
+        }
+      };
+
+  private static final Function<? super Long, Tag> longAttributeConverter =
+      new Function<Long, Tag>() {
+        @Override
+        public Tag apply(final Long value) {
+          final Tag tag = new Tag();
+          tag.setVType(TagType.LONG);
+          tag.setVLong(value);
+          return tag;
+        }
+      };
+
+  private static final Function<Object, Tag> defaultAttributeConverter =
+      new Function<Object, Tag>() {
+        @Override
+        public Tag apply(final Object value) {
+          final Tag tag = new Tag();
+          tag.setVType(TagType.STRING);
+          tag.setVStr(value.toString());
+          return tag;
+        }
+      };
+
+  // Re-usable buffers to avoid too much memory allocation during conversions.
```

Comment:
When exporting trace ID, span ID and options, these buffers allow us to avoid allocating too much memory for each exported span, however, we still copy to these buffers. Would it be better to move the conversion to the `TraceId`, `SpanId` and `TraceOptions` classes? Or would this be pre-mature optimisations at the cost of polluting their APIs?

Reply:
I understand the performance concerns behind that but I am still not sure how big the impact would be, given that the spans are already sampled and the total occupation is really small. Are you suggesting to expose something like `unsafeGetBytes()` to return the underlying buffer instead of a copied one in `TraceId`, `SpanId` and `TraceOptions`, or adding the similar kind of BufferManager logic to reuse a fixed size of buffer? I am fine with the former one as long as invoker knows what they are doing. The latter one seems to me a little bit pre-mature at least for now, though I do see the possible necessity of doing so in some certain scenarios, e.g. a fast streaming processing system.

/cc @bogdandrutu

Reply:
Again, this may be premature optimisation, so feel free to just reply with "_meh, YAGNI_" 😄.

FWIW, I've run a small benchmark, profiled it, and though none of the below calls stands out as a clear bottleneck or performance killer, the ones appearing at the relative "top" are:

- `AttributeValue#match(Function, Function, Function, Function)`
- `AutoValue_Annotation.getAttributes()` (from `annotationEventsToLogs`)
- `TraceId.copyBytesTo(byte[], int)`
- `Ints.fromByteArray(byte[])` (from `optionsToFlags`)
- `Lists.newArrayListWithExpectedSize(int)` (from various calls)

so I guess my intuition wasn't completely off.
Note that the benchmark was specifically on the export, not the entire lifecycle of an application with tracing enabled: this might not matter in the grand scheme of things.

Assuming users sample with low probability, it is probably not important.
However, given I'm new to all this, can we assume that?

In any case, if we end up caring, I can think of the following solutions:
1. Keep it as it is.
    - **Pro**: doesn't pollute the API ; "_OK_" performance as per benchmark.
    - **Con**: array-copies on each sample exported, but "_meh_".
2. Expose something like `unsafeGetBytes`.
    - **Pro**: allow exporter code to get the bytes w/o a copy. 
    - **Con**: pollutes the API a bit.
3. Expose something like `idAsJaegerLong()` (and for `TraceId`: `idAsJaegerHighLong()`/`idAsJaegerLowLong()`).
    - **Pro**: safer than 2. 
    - **Con**: pollutes the API a lot more than 2 ; opens the door to more pollution for all the other exporters.

so I'd probably for go 2. as well @HailongWen.
But I'll defer that call to you and  @bogdandrutu! 🙂
I'm not sure I understand the alternative with the "*BufferManager*" though, sorry.

Reply:
About the `BufferManager`: never mind, it is a little bit off the original topic. I was thinking if you want to move the buffer logic you wrote in the exporter to OpenCensus library as a util or something. It turns out it is not the case.

About sampling: yes, usually the sampling ratio would be small but it all depends on users' scenario. `alwaysSample` can be used sometimes, e.g. if users don't want to miss a rare exception and can afford the overheads. In such case, I will be more concerned about exportation I/O.

About the benchmark: thanks for adding that. I am interested in the details, e.g. what the program is like, what is measured (time? memory? I/O?), what percentage of resources these functions consume.

Reply:
Thanks for the clarifications.

Regarding the benchmark, I used [YourKit](https://www.yourkit.com/) and measured CPU time, while running a modified version of `JaegerExporterHandlerTest`, exporting many times, re-using the same span (which means it is probably cached):
```java
    for (int i = 0; i < 5; i++) {
      handler.export(Collections.nCopies(1000000, spanData));
    }
```
Results aren't exactly identical when re-running it today, but still similar-ish:

![image](https://user-images.githubusercontent.com/836014/37081041-2438c420-21e0-11e8-92f5-6712fef39ca6.png)